kubermatic/kubermatic
# Add OIDC support to the K8S Dashboard integration

**What does this PR do / Why do we need it**:
This PR changes the way Kubernetes Dashboard proxy integration works. It is now based on the OIDC flow and the authorization token returned from the identity provider. Thanks to the KKP Dashboard RBAC management it is now possible to dynamically configure per-user privileges inside K8S Dashboard.

**NOTE**: When this gets merged it will effectively break the `Open Dashboard` button in the KKP UI until other issues/PRs related to this epic will be done.

##### What's Changed
- Removed old dashboard proxy endpoints (ref: #10014):
  - `/api/v2/projects/{project_id}/clusters/{cluster_id}/dashboard/proxy`
  - `/api/v1/projects/{project_id}/clusters/{cluster_id}/dashboard/proxy`
- Added new endpoints to handle proxy calls
  - `api/v2/dashboard/login` - Handles the OIDC redirect and callback parts. Required parameters:
    - `projectID` and `clusterID` - used for the initial request to this endpoint
    - `state` and `code` - those are returned by the identity provider after successful authentication
  - `api/v2/projects/{project_id}/clusters/{cluster_id}/dashboard/proxy` - Required parameters:
    - `token` - bearer token retrieved from the identity provider after successful authentication
- Updated KKP - K8S Dashboard version compatibility matrix

##### Preview
https://user-images.githubusercontent.com/2285385/173351067-aeaf9695-47b4-4411-bebd-312329590936.mp4

**Does this PR close any issues?**:
Related to #9804
Fixes #10016
Fixes #10014

**Special notes for your reviewer**:
In order to test this locally:
1. Alter our dex configuration (`kubermaticIssuer` client) to whitelist the `redirectURI`: `http://localhost:8080/api/v2/dashboard/login`
  - `kubectl -n oauth edit configmap dex`
  - `kubectl -n oauth rollout restart deployment dex`
![image](https://user-images.githubusercontent.com/2285385/173568073-c5a97665-28e0-4d55-80c6-f5133a091a12.png)
3. Create and wait for any cluster to start (using our dev env).
4. Run the API locally, i.e. `make clean && FEATURE_GATES=OIDCKubeCfgEndpoint=true KUBERMATIC_EDITION=ee KUBERMATIC_DEBUG=false ./hack/run-api.sh`
5. There are 2 ways to test the "frontend" part now:
  - run the KKP dashboard using `npm run start:local` after checking out this PR: https://github.com/kubermatic/dashboard/pull/4585
  - go to the `http://localhost:8080/api/v2/dashboard/login?clusterID=<clusterID>&projectID=<projectID>`. Make sure to replace `clusterID` and `projectID` with values of the created cluster. You will also need to get the authorization bearer token and inject it into the request to the API (i.e. using requestly chrome extension).

**NOTE**: The first step will not be required once #10015 will be done.

**Does this PR introduce a user-facing change?**:

```release-note
Removed the `/api/v1/projects/{project_id}/clusters/{cluster_id}/dashboard/proxy` endpoint.
Updated the `/api/v2/projects/{project_id}/clusters/{cluster_id}/dashboard/proxy` to use the OIDC based authentication flow to access K8S Dashboard.
Added the `api/v2/dashboard/login` endpoint to initiate the OIDC login flow.
```
